File: DynamicProgramming/EditDis/edit.cpp
#include <string>
#include <vector>
#include <climits>
using namespace std;

class Solution {
public:
    int doDP(string word1, string word2, int i, int j,vector<vector<int>>& dp){
        if(i==word1.size())return word2.size()-j;
        else if(j==word2.size())return word1.size()-i;
        else if(dp[i][j]!=-1)return dp[i][j];
        else if(word1[i]==word2[j])return dp[i][j] = doDP(word1,word2,i+1,j+1,dp);
        int ans = INT_MAX;
        ans = min(ans,doDP(word1,word2,i+1,j+1,dp)); // Replace
        ans = min(ans,doDP(word1,word2,i,j+1,dp)); // Insert
        ans = min(ans,doDP(word1,word2,i+1,j,dp)); // Delete
        ans += 1;
        dp[i][j] = ans;
        return dp[i][j];
    }
    int minDistance(string word1, string word2) {
        vector<vector<int>> dp(word1.size()+1,vector<int>(word2.size()+1, -1));
        return doDP(word1,word2,0,0,dp);
    }
};

File: DynamicProgramming/EditDis/edit.py
class Solution:
    def editDistance(self, s1, s2):
        n = len(s1)
        m = len(s2)

        dp = [[0 for _ in range(m + 1)] for _ in range(n + 1)]

        # Base cases
        for j in range(m + 1):
            dp[n][j] = m - j

        for i in range(n + 1):
            dp[i][m] = n - i

        # Fill DP table
        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                if s1[i] == s2[j]:
                    dp[i][j] = dp[i + 1][j + 1]
                else:
                    dp[i][j] = 1 + min(
                        dp[i + 1][j + 1],  # Replace
                        dp[i][j + 1],      # Insert
                        dp[i + 1][j]       # Delete
                    )

        return dp[0][0]